src/dotVariant.Generator/DiagnosedResult.cs

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright Miro Knejp 2021.
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE.txt or copy at https://www.boost.org/LICENSE_1_0.txt)
+//
+
+using Microsoft.CodeAnalysis;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace dotVariant.Generator;
+
+/// <summary>A result type carrying <see cref="Diagnostic"/> values and a <typeparamref name="TValue"/> if any only if no errors are diagnosed.</summary>
+/// <typeparam name="TValue">The type of the value</typeparam>
+/// <remarks>Diagnosed result ignores diagnostics during equality comparison to improve caching consistency.</remarks>
+public readonly struct DiagnosedResult<TValue> : IEquatable<DiagnosedResult<TValue>>
+    where TValue: IEquatable<TValue>
+{
+    private readonly bool _noErrors;
```

Comment:
I think it would be better to rename this to `_hasErrors` as that would get rid of three negations and add one, and make it clear it's the backing field for `HasErrors`.

However, this would flip the `default(DiagnosedResult<TValue>).HasError` value from `true` to `false`. Would that cause problems?

Reply:
Sadly, that would violate the struct `default` initializer safety: A struct must be in a valid state when initialized to `default`.
In essence, yes, it would cause problems when returning `default` as an empty error result to be removed.

Reply:
If this is such a critical property of the struct then please add a test to enshrine it for future contributors. This is something I can easily see someone trying to "cleanup".